### PR TITLE
A binary search algorithm for finding opt sf

### PR DESCRIPTION
**Pull request by Vimalakirti**
This PR added the following features:
- implemented a binary search algorithm for finding the optimal scale factor.
- to achieve the above goal, also added `run` for CQ and CQ2 to check their ranges.
- (the editions in `softmax.rs` and `onnx.rs` are just re-formatting)

### CODE REVIEW

**Review by liliatangxy — DISMISSED** [review]

File: src/bin/search_sf.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+use core::panic;
+use std::fs;
+use std::process::{Command, Stdio};
+use zk_torch::{CONFIG, CONFIG_FILE};
+
+// Run the witness_gen binary and return the std output
+fn witness_gen() -> String {
+  let output = Command::new("./target/release/witness_gen")
```
Comment (ddkang) [inline src/bin/search_sf.rs]: Don't use these random strings

File: src/bin/search_sf.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  // Capture stderr (where panic messages typically appear)
+  let stderr = String::from_utf8_lossy(&output.stderr);
+  if !stderr.is_empty() {
+    println!("stdout msg:\n{}", stdout);
+    println!("stderr msg:\n{}", stderr);
+  }
+  format!("{}", stdout)
+}
+
+fn modify_config_sf(current_sf: usize, file_path: &str) {
```
Comment (ddkang) [inline src/bin/search_sf.rs]: This seems like a really bad way to do this

Comment (ddkang) [reply src/bin/search_sf.rs]: Is this a global SF?

File: src/bin/search_sf.rs (hunk 1 above)
Comment (Vimalakirti) [reply src/bin/search_sf.rs]: Removed

**Comment by Vimalakirti** [comment]
The implementation allows us to search the optimal SF by running witness generation multiple times with different scale factors using the binary `search_sf`. Besides, the interfaces of functions in `search_sf.rs`  are now easier to understand. I recommend to run it with real/fixed input instead of random input (warning will pop out) because the optimal sf might fail when using another random input.

Specifically, the new commit catches the "CQ out of range error" by making `basicblock.run` and `graph.run` return `Result`. It also used a `RwLock` to wrap the `SF_LOG`, `SF`, and `SF_FLOAT` so that we can reset their values after initialization.

Please ignore all the edited files under `basic_block/` and `layer` as their functionalities are not changed.

**Review by liliatangxy — APPROVED** [review]

**Review by ddkang — APPROVED** [review]

**Pull request merged**